File: following_patterns.js


function areFollowingPatterns(strings, patterns) {
    if( strings.length != patterns.length){
        return false
    }
    let hashPS = {};
    let hashSP = {};
    for (let i = 0; i< strings.length; i++){
        let currentPattern = patterns[i]
        let currentString = strings[i]
        if(!hashPS[currentPattern]) {
            hashPS[currentPattern] = currentString
        } else if (hashPS[currentPattern] != currentString){
            return false
        }
        if(!hashSP[currentString]) {
            hashSP[currentString] = currentPattern
        } else if(hashSP[currentString] != currentPattern){
            return false
        }
    }
    return true
}
string = ["cat", "dog", "doggy"]
pattern = ["a", "b", "b"]

console.log(areFollowingPatterns(string, pattern))

File: following_patterns.py
def areFollowingPatterns(string, pattern):
    string_set = set()
    pattern_set = set()

    for i in string:
        string_set.add(i)
    for i in pattern:
        pattern_set.add(i)
    return len(string_set) == len(pattern_set)


string = ["cat", "dog", "doggy"]
pattern = ["a", "b", "b"]

print(areFollowingPatterns(string, pattern))
